File: Pascal/Shared/CastleAppUnit.pas
unit CastleAppUnit;

interface

uses  System.SysUtils, System.Classes, System.Types, CastleViewport, CastleUIControls, CastleScene, CastleVectors, CastleTransform;

type
  { TCastleSceneHelper }
  TCastleSceneHelper = class helper for TCastleScene
    function Normalize: Boolean;
    { Fit the Scene in a 1x1x1 box }
  end;

  { TCastleCameraHelper }
  TCastleCameraHelper = class helper for TCastleCamera
    procedure ViewFromRadius(const ARadius: Single; const ACamPos: TVector3);
    { Position Camera ARadius from Origin pointing at Origin }
  end;

  { TCastleApp }
  TCastleApp = class(TCastleView)
    procedure Update(const SecondsPassed: Single; var HandleInput: Boolean); override; // TCastleUserInterface
    procedure Start; override; // TCastleView
    procedure Stop; override; // TCastleView
    procedure Resize; override; // TCastleUserInterface
  private
    Camera: TCastleCamera;
    CameraLight: TCastleDirectionalLight;
    Viewport: TCastleViewport;
    function CreateDirectionalLight(LightPos: TVector3): TCastleDirectionalLight;
    function LoadScene(filename: String): TCastleScene;
    procedure LoadViewport;
  public
    ActiveScene: TCastleScene;
    constructor Create(AOwner: TComponent); override;
    destructor Destroy; override;
    procedure SwitchView3D(const Use3D: Boolean);
    procedure SetRotation(const ARotDeg: Single);
  end;


implementation

uses Math, CastleProjection, CastleFilesUtils;

function TCastleSceneHelper.Normalize: Boolean;
var
  BBMax: Single;
begin
  Result := False;
  if not(RootNode = nil) then
    begin
    if not LocalBoundingBox.IsEmptyOrZero then
      begin
        if LocalBoundingBox.MaxSize > 0 then
          begin
            Center := Vector3(Min(LocalBoundingBox.Data[0].X, LocalBoundingBox.Data[1].X) + (LocalBoundingBox.SizeX / 2),
                              Min(LocalBoundingBox.Data[0].Y, LocalBoundingBox.Data[1].Y) + (LocalBoundingBox.SizeY / 2),
                              Min(LocalBoundingBox.Data[0].Z, LocalBoundingBox.Data[1].Z) + (LocalBoundingBox.SizeZ / 2));
            Translation := -Center;

            BBMax := 1 / LocalBoundingBox.MaxSize;
            Scale := Vector3(BBMax,
                             BBMax,
                             BBMax);
            Result := True;
          end;
      end;
    end;
end;

constructor TCastleApp.Create(AOwner: TComponent);
begin
  inherited Create(AOwner);
end;

destructor TCastleApp.Destroy;
begin
  inherited;
end;

procedure TCastleApp.Update(const SecondsPassed: Single; var HandleInput: Boolean);
begin
  inherited;
end;

procedure TCastleApp.Resize;
begin
  Viewport.Width := Container.UnscaledWidth;
  Viewport.Height := Container.UnscaledHeight;
  if Camera.ProjectionType = ptOrthographic then
    begin
      if Viewport.Width > Viewport.Height then
        Camera.Orthographic.Height := 1
      else
        Camera.Orthographic.Width := 1;
    end;
end;

procedure TCastleApp.SetRotation(const ARotDeg: Single);
begin
   ActiveScene.Rotation := Vector4(0, 1, 0,  ARotDeg * Pi / 180);
end;

procedure TCastleApp.Start;
var
 datadir: String;
begin
  inherited;
  LoadViewport;
  ActiveScene := LoadScene('castle-data:/KayKit/KayKit_Adventurers_1.0_FREE/Characters/gltf/Knight.glb');

  if Assigned(ActiveScene) then
    begin
      ActiveScene.Normalize;
      Viewport.Items.Add(ActiveScene);
    end;
end;

procedure TCastleApp.Stop;
begin
  inherited;
end;

function TCastleApp.CreateDirectionalLight(LightPos: TVector3): TCastleDirectionalLight;
var
  Light: TCastleDirectionalLight;
begin
  Light := TCastleDirectionalLight.Create(Self);

  Light.Direction := LightPos;
  Light.Color := Vector3(1, 1, 1);
  Light.Intensity := 1;

  Result := Light;
end;

procedure TCastleApp.LoadViewport;
begin
  Viewport := TCastleViewport.Create(Self);
  Viewport.FullSize := False;
  Viewport.Width := Container.UnscaledWidth;
  Viewport.Height := Container.UnscaledHeight;
  Viewport.Transparent := True;

  Camera := TCastleCamera.Create(Viewport);

  Camera.ViewFromRadius(2, Vector3(1, 1, 1));

  CameraLight := CreateDirectionalLight(Vector3(0,0,1));
  Camera.Add(CameraLight);

  Viewport.Items.Add(Camera);
  Viewport.Camera := Camera;

  InsertFront(Viewport);
end;

procedure TCastleApp.SwitchView3D(const Use3D: Boolean);
begin
  if Use3D then
    begin
      Camera.ProjectionType := ptPerspective;
      Camera.ViewFromRadius(2, Vector3(1, 1, 1));
    end
  else
    begin
      Viewport.Setup2D;
      Camera.ProjectionType := ptOrthographic;
      Camera.Orthographic.Width := 1;
      Camera.Orthographic.Origin := Vector2(0.5, 0.5);
    end;
  Resize;
end;

function TCastleApp.LoadScene(filename: String): TCastleScene;
begin
  Result := Nil;
  try
    Result := TCastleScene.Create(Self);
    Result.Load(filename);
  except
    on E : Exception do
      begin
        Raise Exception.Create('Error in LoadScene : ' + E.ClassName + ' - ' + E.Message);
       end;
  end;
end;

procedure TCastleCameraHelper.ViewFromRadius(const ARadius: Single; const ACamPos: TVector3);
var
  Spherical: TVector3;
begin
  Spherical := ACamPos.Normalize;
  Spherical := Spherical * ARadius;
  Up := Vector3(0, 1, 0);
  Direction := -ACamPos;
  Translation  := Spherical;
end;


end.

File: CPP/Shared/CastleAppUnit.cpp
//---------------------------------------------------------------------------

#pragma hdrstop

#include "CastleAppUnit.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)

__fastcall TCastleApp::TCastleApp(TComponent* Owner) : TCastleView(Owner)
{
}

__fastcall TCastleApp::~TCastleApp()
{
}

__fastcall void TCastleApp::Start()
{
	TCastleView::Start();
	LoadViewport();
	ActiveScene = LoadScene("castle-data:/KayKit/KayKit_Adventurers_1.0_FREE/Characters/gltf/Knight.glb");
	Viewport->Items->Add(ActiveScene);
}

__fastcall void TCastleApp::Stop()
{
	TCastleView::Stop();
};

__fastcall void TCastleApp::Resize()
{
	TCastleUserInterface::Resize();

	Viewport->Width = Container()->UnscaledWidth();
	Viewport->Height = Container()->UnscaledHeight();

	if (Camera->ProjectionType == ptOrthographic)
	{
		if (Viewport->Width > Viewport->Height)
		{
			Camera->Orthographic->Height = 2;
		}
		else
		{
			Camera->Orthographic->Width = 2;
		}
	}
};

__fastcall void TCastleApp::Update(const float SecondsPassed, bool &HandleInput)
{
	TCastleView::Update(SecondsPassed, HandleInput);
}

__fastcall void TCastleApp::LoadViewport()
{
	TVector3 ALightPos;
	ALightPos = Vector3(0,0,1);
	Viewport = new TCastleViewport(this);
	Viewport->FullSize = False;
	Viewport->Width = Container()->UnscaledWidth();
	Viewport->Height = Container()->UnscaledHeight();
	Viewport->Transparent = True;

	Camera = new TCastleCamera(Viewport);

	ViewFromRadius(2, Vector3(1, 1, 1));

	CameraLight = CreateDirectionalLight(ALightPos);

	Camera->Add(CameraLight);

	Viewport->Items->Add(Camera);
	Viewport->Camera = Camera;

	InsertFront(Viewport);

};

__fastcall TCastleDirectionalLight* TCastleApp::CreateDirectionalLight(TVector3 LightPos)
{
	TCastleDirectionalLight *Light = new TCastleDirectionalLight(this);
	Light->Direction = LightPos;
	Light->Color = Vector3(1, 1, 1);
	Light->Intensity = 1;

	return Light;
};

__fastcall void  TCastleApp::SwitchView3D(const bool Use3D)
{
  if(Use3D)
  {
	  Camera->ProjectionType = ptPerspective;
	  ViewFromRadius(2, Vector3(1, 1, 1));
  }
  else
  {
	  Viewport->Setup2D();
	  Camera->ProjectionType = ptOrthographic;
	  Camera->Orthographic->Origin = Castlevectors::Vector2(0.5, 0.5);
  }
  Resize();
};

__fastcall TCastleScene* TCastleApp::LoadScene(String filename)
{
	TCastleScene* NewScene = new TCastleScene(this);
	NewScene->Load(filename);
	NormalizeScene(NewScene);
	return NewScene;
}

__fastcall void TCastleApp::ViewFromRadius(const Single ARadius, const TVector3 ACamPos)
{
  TVector3 Spherical;

  Spherical = ACamPos;
  Spherical.Normalize();
  Spherical = Spherical * ARadius;
  Camera->Up = Vector3(0, 1, 0);
  Camera->Direction = Vector3(-ACamPos.X, -ACamPos.Y, -ACamPos.Z);
  Camera->Translation = Spherical;
};

__fastcall void TCastleApp::NormalizeScene(TCastleScene *Scene)
{
  Single BBMax;
  TBox3D Box;

  if(Scene->RootNode != nullptr)
  {
	Box = Scene->LocalBoundingBox();
	if(!Box.IsEmptyOrZero())
	{
		if(Box.MaxSize() > 0)
		{
			Scene->Center = Vector3(  Min(Box.Data[0].X, Box.Data[1].X) + (Box.SizeX() / 2),
									  Min(Box.Data[0].Y, Box.Data[1].Y) + (Box.SizeY() / 2),
									  Min(Box.Data[0].Z, Box.Data[1].Z) + (Box.SizeZ() / 2));
			Scene->Translation = Vector3(-Scene->Center.X, -Scene->Center.Y, -Scene->Center.Z);
			BBMax = 2 / Box.MaxSize();
			Scene->Scale = Vector3(BBMax, BBMax, BBMax);
		}
	}
  }
}

__fastcall void TCastleApp::SetRotation(const float ARotDeg)
{
  if(ActiveScene != nullptr)
  {
	  ActiveScene->Rotation = Vector4(0, 1, 0,  ARotDeg * M_PI / 180);
  }
};

